### ISSUE TITLE
Pyright incorrectly computes whether a generator is async

### ISSUE THREAD

**Issue by mniip** [microsoft/pyright#6999]
**Describe the bug**

A generator of the form `f(x) for x in await a` does not in fact create an `async_generator`. Rather it first awaits `a` and then creates a regular `generator`.

**Code or Screenshots**
```py
from typing import Awaitable
async def foo(xs: Awaitable[list[int]]) -> list[int]:
    return list(x for x in await xs)
```
```
  a.py:3:17 - error: Argument of type "AsyncGenerator[int, None]" cannot be assigned to parameter "__iterable" of type "Iterable[_T@list]" in function "__init__"
    "AsyncGenerator[int, None]" is incompatible with protocol "Iterable[_T@list]"
      "__iter__" is not present (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
Reproducible on pyright 1.1.347
Not reproducible on pyright 1.1.345

**Comment by erictraut** [comment]
I made this change in response to [this bug report](https://github.com/microsoft/pyright/issues/6948).

The [Python documentation](https://docs.python.org/3/reference/expressions.html#generator-expressions) says:

> If a comprehension contains either `async for` clauses or `await` expressions or other asynchronous comprehensions it is called an _asynchronous comprehension_.

It also says:

> If a generator expression contains either async for clauses or [await](https://docs.python.org/3/reference/expressions.html#await) expressions it is called an asynchronous generator expression. An asynchronous generator expression returns a new asynchronous generator object, which is an asynchronous iterator (see [Asynchronous Iterators](https://docs.python.org/3/reference/datamodel.html#async-iterators)).

So I'm confused why the above wouldn't generate an async generator. Am I misinterpreting the documentation? Are the docs wrong?

**Comment by mniip** [comment]
Yeah it seems that [PEP 530](https://peps.python.org/pep-0530/#await-in-comprehensions) doesn't actually specify, merely lists some examples. I cannot tell whether it's the cpython implementation or the cpython docs that are wrong, but they definitely do not match:
```py
import asyncio
async def foo():
    print("Entered foo")
    return [1, 2, 3]
async def main():
    gen = (x for x in await foo())
    print(gen)
    print(list(gen))
asyncio.run(main())
```
```
Entered foo
<generator object main.<locals>.<genexpr> at 0x7f856b418110>
[1, 2, 3]
```
This is in contrast to when an `await` appears in a `for` clause that is not the first:
```py
import asyncio
async def foo():
    print("Entered foo")
    return [1, 2, 3]
async def main():
    gen = (x for _ in () for x in await foo())
    print(gen)
    print(list(gen))
asyncio.run(main())
```
```py
<async_generator object main.<locals>.<genexpr> at 0x7ff441e2cd60>
Traceback (most recent call last): ...
TypeError: 'async_generator' object is not iterable
```

**Comment by mniip** [comment]
Maybe this line in the [docs](https://docs.python.org/3/reference/expressions.html#generator-expressions) is a clue:

> However, the iterable expression in the leftmost for clause is immediately evaluated, so that an error produced by it will be emitted at the point where the generator expression is defined, rather than at the point where the first value is retrieved.

Still it's not 100% clear what "evaluated" means for an await expression.

**Comment by mniip** [comment]
I've created https://github.com/python/cpython/issues/114104

**Comment by erictraut** [comment]
Thanks for helping track down the confusion in the documentation. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.348, which was just published. It will also be included in a future release of pylance.
